File: eli5/src/ai/flows/cartoonify.ts
import { ai } from "../genkit.js";
import { googleAI } from "@genkit-ai/google-genai";
import { z } from "genkit";

const CartoonifyRequestSchema = z.object({
  image: z
    .string()
    .describe("A data URI of an image of a person to cartoonify"),
});

type CartoonifyRequest = z.infer<typeof CartoonifyRequestSchema>;

export const cartoonify = ai.defineFlow(
  {
    name: "cartoonify",
    inputSchema: CartoonifyRequestSchema,
    outputSchema: z.string(),
  },
  async ({ image }: CartoonifyRequest) => {
    const { media } = await ai.generate({
      model: googleAI.model("gemini-2.5-flash-image-preview"),
      config: { responseModalities: ["TEXT", "IMAGE"] },
      prompt: [
        {
          text: "Transform the person in the following image into a full-body cartoon character in a neutral pose. The background should be white.",
        },
        { media: { url: image } },
      ],
    });
    if (!media) {
      throw new Error("Image generation failed.");
    }
    return media.url;
  },
);

File: eli5_dart/lib/src/flows/cartoonify.dart
import 'package:genkit/genkit.dart';
import 'package:genkit_google_genai/genkit_google_genai.dart';
import 'package:schemantic/schemantic.dart';

import '../genkit.dart';

part 'cartoonify.g.dart';

@Schema()
abstract class $CartoonifyRequest {
  String get image;
}

final cartoonifyFlow = ai.defineFlow(
  name: 'cartoonify',
  inputSchema: CartoonifyRequest.$schema,
  outputSchema: SchemanticType.string(),
  fn: (CartoonifyRequest input, _) async {
    final response = await ai.generate(
      model: googleAI.gemini('gemini-2.5-flash-image'),
      messages: [
        Message(
          role: Role.user,
          content: [
            TextPart(text: 'Transform the person in the following image into a full-body cartoon character in a neutral pose. The background should be white.'),
            MediaPart(media: Media(url: input.image)),
          ],
        ),
      ],
    );
    
    final mediaUrl = response.media?.url;
    if (mediaUrl == null) {
      throw Exception('Image generation failed.');
    }
    return mediaUrl;
  },
);
